app/src/main/kotlin/model/io/neo4j/Neo4jRepository.kt

```diff
@@ -0,0 +1,182 @@
+
+        val isDirected = hasDirection(graphContents[0])
+        val isWeighted = hasWeight(graphContents[0])
+
+        val graph = initializeGraph(isWeighted, isDirected)
+
+        val graphSize = getStoredGraphSize(name)
+
+        val wasVertexLoaded = MutableList(graphSize) { false }
+        val addedVerticesList = MutableList(graphSize) { Vertex(-1, "-1") }
+
+        for (content in graphContents) {
+            val id1 = content["id1"].asInt()
+            val data1 = content["data1"].toString()
+
+            val id2 = content["id2"].asInt()
+            val data2 = content["data2"].toString()
+
+            val weightString = content["weight"].toString()
+            val edgeWeight = if (weightString != "NULL") weightString.toInt() else null
+
+            if (!wasVertexLoaded[id1]) {
+                wasVertexLoaded[id1] = true
+                addedVerticesList[id1] = graph.addVertex(data1)
+            }
+            if (!wasVertexLoaded[id2]) {
+                wasVertexLoaded[id2] = true
+                addedVerticesList[id2] = graph.addVertex(data2)
+            }
+
+            val v1 = addedVerticesList[id1]
+            val v2 = addedVerticesList[id2]
+
+            if (edgeWeight == null) {
+                graph.addEdge(v1, v2)
+            } else if (isDirected) {
+                graph as WeightedDirectedGraph
+                graph.addEdge(v1, v2, edgeWeight)
+            } else {
+                graph as WeightedUndirectedGraph
+                graph.addEdge(v1, v2, edgeWeight)
+            }
+        }
+
+        return graph
+    }
+
+    private fun initializeGraph(isWeighted: Boolean, isDirected: Boolean): Graph<String> {
+        return if (isWeighted && isDirected) WeightedDirectedGraph()
+        else if (isWeighted) WeightedUndirectedGraph()
+        else if (isDirected) DirectedGraph()
+        else UndirectedGraph()
+    }
+
+    private fun readGraphContents(name: String) = session.executeRead { tx ->
+        tx.run(
+            "MATCH (v:$name)-[e]->(u:$name) " +
+            "RETURN " +
+            "v.id AS id1, " +
+            "v.data AS data1, " +
+            "u.id AS id2, " +
+            "u.data AS data2, " +
+            "e.weight AS weight, " +
+            "type(e) AS relationType "
+        ).list()
+    }
+
+    private fun getStoredGraphSize(name: String): Int {
+        val result = session.executeRead { tx ->
+            tx.run(
+                "MATCH (v:$name) " +
+                "RETURN count(v) AS size "
+            ).list()
+        }
+
+        val size = result[0]["size"].asInt()
+
+        return size
+    }
+
+    private fun hasWeight(graphRecord: Record): Boolean {
+        return graphRecord["weight"].toString() != "NULL"
+    }
+
+    private fun hasDirection(graphRecord: Record): Boolean {
+        return graphRecord["relationType"].asString() == DIR_LABEL
+    }
+
+    override fun close() {
+        session.close()
+        driver.close()
+    }
```

Comment:
❗️ Does this ever get called? It looks like once it has been used it stays opened forever which is not right. The right way would probably be to close everything after the graph has been saved/loaded. This way you won\t need `Neo4jRepositoryHandler` as a global singleton since the database connection would be opened and closed locally in the code that handles saving/loading. If you want to let the user to load/save multiple graphs without re-entering their credentials, you should either just cache the credentials or leave the connection open but find some other way to close it at some moment.